standalone/pruntime/enclave/src/system/gk.rs

```diff
@@ -122,7 +117,44 @@ where
         self.egress.set_dummy(true);
     }
 
+    pub fn master_pubkey_uploaded(&mut self) {
+        self.master_pubkey_on_chain = true;
+    }
+
+    pub fn share_master_key(&mut self, pubkey: &WorkerPublicKey, ecdh_pubkey: &EcdhPublicKey) {
+        info!("Gatekeeper: try dispatch master key");
+        let derived_key = self
+            .master_key
+            .derive_sr25519_pair(&[&crate::generate_random_info()])
+            .expect("should not fail with valid info; qed.");
+        let my_ecdh_key = derived_key
+            .derive_ecdh_key()
+            .expect("ecdh key derivation should never failed with valid master key; qed.");
+        let secret = ecdh::agree(&my_ecdh_key, &ecdh_pubkey.0)
+            .expect("should never fail with valid ecdh key; qed.");
+        let iv = crate::generate_random_iv();
+        let mut data = self.master_key.dump_secret_key().to_vec();
+
+        aead::encrypt(&iv, &secret, &mut data).expect("Failed to encrypt master key");
+        self.egress
+            .push_message(MasterKeyEvent::dispatch_master_key_event(
+                pubkey.clone(),
+                my_ecdh_key
+                    .public()
+                    .as_ref()
+                    .try_into()
+                    .expect("should never fail given pubkey with correct length; qed;"),
+                data,
+                iv,
+            ));
+    }
+
     pub fn process_messages(&mut self, block: &BlockInfo<'_>) {
+        if !self.master_pubkey_on_chain {
+            info!("Gatekeeper: not handle messages for no master pubkey on chain");
```

Comment:
```suggestion
            info!("Gatekeeper: not handling the messages because Gatekeepers are not launched on chain");
```

Let's use Gatekeeper "launch" to describe the state where the master pubkey is already generated and stored on the blockchain.